### ISSUE TITLE
Add ttlSecondsUntilTainted for better node lifecycle control

### ISSUE THREAD

**Issue by stevehipwell** [aws/karpenter#1526]
**Tell us about your request**
I'd like the ability to taint a node after a given period via `ttlSecondsUntilTainted` to allow running pods to finish before the node is terminated, this should still respect the `ttlSecondsUntilExpired` and `ttlSecondsAfterEmpty`.

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
When running jobs such as CI/CD pipelines in K8s there are long running jobs which shouldn't be terminated due to the high cost of re-running them, by adding `ttlSecondsUntilTainted` we can have nodes that expire and are replaced without the cost of killing potentially long running jobs.

**Are you currently working around this issue?**
Not using Karpenter.

**Additional context**
n/a

**Attachments**
n/a

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by ellistarn** [comment]
Does annotating your longrunning jobs with `karpenter.sh/do-not-evict` work for you?

**Comment by stevehipwell** [comment]
@ellistarn wouldn't this allow more jobs to schedule onto the node and basically make it unterminatable?

**Comment by ellistarn** [comment]
The node will still expire and cordon, but it won't start draining until the pod with those annotations are done. @njtran can you confirm?

**Comment by njtran** [comment]
Yep! If you have a pod with a `do-not-evict` annotation, Karpenter will not drain any of the pods on a node. 

A node is not drained if it has a `do-not-evict` pod, and we will not terminate a node until that node is considered drained:
- https://github.com/aws/karpenter/blob/bb65375ae4f1a6ca1a4a30f217162abf71bd5dcc/pkg/controllers/termination/controller.go#L85-L91
- https://github.com/aws/karpenter/blob/bb65375ae4f1a6ca1a4a30f217162abf71bd5dcc/pkg/controllers/termination/terminate.go#L68-L70

**Comment by stevehipwell** [comment]
@njtran this fits my use case happy path perfectly, but it opens up a separate concern that the `do-not-evict` annotation could keep a node running indefinitely? If there isn't a way to do this already I'm happy to close this issue as being a "didn't read the docs correctly" and open a new issue regarding a `do-not-evict` grace period?

**Comment by stevehipwell** [comment]
@njtran fundamentally it comes down to reducing the blast radius. If anyone who can start a container can keep a node running that starts to look like a potential problem. I'd like to know that after a given time, no matter what a user has done, that a node will terminate and be replaced.

**Comment by ellistarn** [comment]
> additional grace period for a do-not-evict

Is there a difference between this and `terminationGracePeriodSeconds`?

**Comment by tzneal** [comment]
> I'd like to know that after a given time, no matter what a user has done, that a node will terminate and be replaced.

Even if it would violate a pod disruption budget?

**Comment by stevehipwell** [comment]
> Is there a difference between this and terminationGracePeriodSeconds?

@ellistarn this would terminate the node even if there was a pod with a `karpenter.sh/do-not-evict` annotation running. Effectively after the TTL expires the annotation would be ignored.

> Even if it would violate a pod disruption budget?

@tzneal I think that's a separate concern, but pretty important. A TTL which guarantees the node is terminated would make sense, having pods stuck due to user error and blocking node maintenance is a common problem when running a platform.

**Comment by njtran** [comment]
Renamed this issue to be more accurate of the discussion above

**Comment by alekhrycaiko** [comment]
> Yep! If you have a pod with a do-not-evict annotation, Karpenter will not drain any of the pods on a node, but it will still cordon the node.

@njtran Curious if this is still an expected behaviour. When enabling consolidation, I experienced Nodes being cordoned but not drained. I'm wondering if this annotation on a Pod is what's possibly happening. 

> A TTL which guarantees the node is terminated would make sense, having pods stuck due to user error and blocking node maintenance is a common problem when running a platform.

+1 a ttl would be helpful.

**Comment by njtran** [comment]
@alekhrycaiko sorry for the late response here, you should be able to see the kube-events that show if a node was unable to be deprovisioned due to a blocking pod. 

And @stevehipwell sorry for the late response! This makes sense to me, as it essentially guarantees (besides involuntary disruptions) that a workload will be up for at least the duration it's decided to be. This would require some design, and would be great for anyone to take up implementing if they want some more experience in the code.

**Comment by njtran** [comment]
Thinking about it more, is there a reason you can't rely on `activeDeadlineSeconds` for this? https://kubernetes.io/docs/concepts/workloads/controllers/job/#job-termination-and-cleanup

If a pod is terminal, we ignore the do-not-evict pod as well. So is there something other than jobs that you want a `do-not-evict` TTL for?

**Comment by stevehipwell** [comment]
@njtran this functionality request is to taint a node so no further pods are scheduled on it allowing either the running pods to finish or the expired TTL to trigger the pod(s) to be killed. So a use case would be to have a grace period difference between the taint TTL and the expired TTL to support clean node terminations when pod age is within a given range which is lower than the grace period; for example a CI/CD system.

**Comment by njtran** [comment]
Ah, I think I fully understand now. So this TTLSecondsUntilTainted (or maybe better named TTLSecondsUntilCordoned) would act as a programmatic schedulable window for all nodes. If you combine this with `ttlSecondsUntilExpired`, then you get a grace period of time where you can guarantee your workloads won't be expired (although you can't guarantee it for other disruptions). Am I right?

Since the only disruption mechanism that is driven by time is expiration, this TTLUntilCordoned seems more likely to make it harder for Consolidation to help cost, and without any enabled disruption mechanisms, this could result in a lot of unused capacity. I think this would be better driven through at the application level, with using `activeDeadlineSeconds` for jobs with `do-not-evict` or adding a `do-not-evict-ignore-ttl` if you there's a reason for pods that aren't jobs.

**Comment by stevehipwell** [comment]
@njtran for the use case I'm referring to the pods are explicitly not jobs and are expensive to re-run, for example a CI/CD system running E2E tests where terminating the pod would have a high cost. The desired behaviour would be for Karpenter to be able to cordon a node to stop any new pods being scheduled while allowing existing pods to either complete successfully or timeout before the node is replaced. It might be best to drive this as a pod label such as `cordon-timeout-seconds: "120"`. Consolidation could then de-prioritise nodes with pods labeled for this behaviour.

**Comment by njtran** [comment]
My apologies on the length of time on this discussion, I've lost track of this issue many times 😞.

If the pods are explicitly not jobs, where terminating the pod would have a high cost, how do you coordinate when the pod is okay to be evicted? Is there some special semantic for this CI/CD pod that signifies completion to another controller in your cluster that isn't represented by a kubernetes-native construct? 

It seems odd to me to create a node-level time-based restriction on scheduling, which is actually motivated by a pod-level time-based restriction. While you can get into a situation where a node is not optimally utilized from unluckiness with how these pods get rescheduled, if you're eventually expecting the node to be expired, since Karpenter doesn't cordon the node, the node should still be utilized. Once no pods have scheduled to it, then Karpenter could execute expiration. But you're right that if `do-not-evict` pods unluckily always opt out the node of deprovisioning, Karpenter may never deprovision this node. 

In my eyes, given the tight correlation with Consolidation and Expiration here, I'm not convinced that this is something that makes sense to include in the API. I think this would better be modeled as https://github.com/aws/karpenter/issues/2988. Thoughts @stevehipwell ?

**Comment by stevehipwell** [comment]
I think that #2988 would probably solve enough of this use case to be viable.

Architecturally speaking though I'm not 100% convinced that a pod (and therefore potentially a low permissioned user) should be able to impact and effectively own a node lifecycle, as is the case currently and even if #2988 is implemented; this looks like a privilege escalation to me. By providing a mechanism for the cluster admin to set a max node ttl after it's lifecycle should have been ended this issue is addressed.

**Comment by jonathan-innis** [comment]
> should be able to impact and effectively own a node lifecycle

Can you try controlling this with something like Kyverno or some other admission webhook mechanism that would block "normal" users from setting this annotation value on a pod to avoid the privilege escalation?

**Comment by stevehipwell** [comment]
> Can you try controlling this with something like Kyverno or some other admission webhook mechanism that would block "normal" users from setting this annotation value on a pod to avoid the privilege escalation?

I can and likely will run OPA Gatekeeper to control this when we get Karpenter into production; but not everyone who runs Karpenter will have any policy capability.

**Comment by yuvalavidor** [comment]
Hi there, I hope im barging into the right discussion here. if no, I am truly sorry.
I want to give an example of a case to see if that makes any sense.

I have a karpenter provisioner with consolidation disabled, that holds a statefulset with 4 replicas on 4 nodes.
the pods have a do-not-evict annotation because I do not want unwanted evictions if a new AMI is released or changes happen.

all PDB's are in place.
in case i need to upgrade my worker nodes (cluster upgrades or ami upgrades)
a restart to the statefulset will start by terminating the pod, then wait for karptenter to catch the pending pod and and spin up a node. if i could use the drift feature the node will be ready beforehand and expedite the process.

I would love to have a way to "trigger "a drift on nodes so i can replace them gracefully. with an annotation or something of the sort.
is that possible and im missing anything? is that planned?

Karpenter v0.30.0
